python/pylibraft/pylibraft/neighbors/ivf_flat/ivf_flat.pyx

```diff
@@ -0,0 +1,712 @@
+            After training the coarse and fine quantizers, we will populate
+            the index with the dataset if add_data_on_build == True, otherwise
+            the index is left empty, and the extend method can be used
+            to add new vectors to the index.
+        adaptive_centers : bool, default = False
+            By default (adaptive_centers = False), the cluster centers are
+            trained in `ivf_flat::build`, and and never modified in
+            `ivf_flat::extend`. The alternative behavior (adaptive_centers
+            = true) is to update the cluster centers for new data when it is
+            added. In this case, `index.centers()` are always exactly the
+            centroids of the data in the corresponding clusters. The drawback
+            of this behavior is that the centroids depend on the order of
+            adding new data (through the classification of the added data);
+            that is, `index.centers()` "drift" together with the changing
+            distribution of the newly added data.
+        """
+        self.params.n_lists = n_lists
+        self.params.metric = _get_metric(metric)
+        self.params.metric_arg = 0
+        self.params.kmeans_n_iters = kmeans_n_iters
+        self.params.kmeans_trainset_fraction = kmeans_trainset_fraction
+        self.params.add_data_on_build = add_data_on_build
+        self.params.adaptive_centers = adaptive_centers
+
+    @property
+    def n_lists(self):
+        return self.params.n_lists
+
+    @property
+    def metric(self):
+        return self.params.metric
+
+    @property
+    def kmeans_n_iters(self):
+        return self.params.kmeans_n_iters
+
+    @property
+    def kmeans_trainset_fraction(self):
+        return self.params.kmeans_trainset_fraction
+
+    @property
+    def add_data_on_build(self):
+        return self.params.add_data_on_build
+
+    @property
+    def adaptive_centers(self):
+        return self.params.adaptive_centers
+
+
+cdef class Index:
+    # We store a pointer to the index because it dose not have a trivial
+    # constructor.
+    cdef c_ivf_flat.index[float, uint64_t] * index_float
+    cdef c_ivf_flat.index[int8_t, uint64_t] * index_int8
+    cdef c_ivf_flat.index[uint8_t, uint64_t] * index_uint8
+
+    cdef readonly bool trained
+
+    def __cinit__(self, handle=None):
+        self.trained = False
+        if handle is None:
+            handle = DeviceResources()
+        cdef device_resources* handle_ = \
+            <device_resources*><size_t>handle.getHandle()
+
+        self.index_int8 = NULL
+        self.index_uint8 = NULL
+        # We create a placeholder object. The actual parameter values do
+        # not matter, it will be replaced with a built index object later.
+        self.index_float = new c_ivf_flat.index[float, uint64_t](
+            deref(handle_), _get_metric("sqeuclidean"),
+            <uint32_t>1,
+            <uint32_t>False,
+            <uint32_t>8)
```

Comment:
This should be changed, it is probably better to initialize to NULL, and allocate only when needed.